### PR TITLE
Update lisk-codec to disallow decoding additional fields - Closes #7047 

### LINKED ISSUE #7047: Update lisk-codec to disallow additional or missing data while decoding
### Description

Currently, `x == encode(decode(x))` is enforced at [processor](https://github.com/LiskHQ/lisk-sdk/blob/v5.2.1/framework/src/node/processor/processor.ts#L292-L297).
However, for efficiency and consistency, this should be implemented in decode of the codec.

While decoding,
- It should fail if existing field number in the schema does not exist in the binary, except for array. (empty array is allow not to appear in the binary)
- It should fail if additional data exist in the binary regardless of position

Additionally, missing tests in the acceptance criteria should be implemented.

### Acceptance Criteria
- Decoding sint... values (using readSInt...) that are out of range. Expected: Decoding should fail
- Decoding a boolean value where the value is a byte unequal to 0x00 and 0x01. Expected:  Decoding should fail
- Encoding a string that contains some non-ASCII characters. Expected:  Encoding succeeds
- Decoding a string that contains some non-ASCII characters. Expected:  Decoding succeeds
- Encoding empty array, where the datatype of the items implies non-packed encoding, e.g. string. Expected:  binary message is empty
- Schema contains a property of type array, where the datatype of the items implies non-packed encoding, e.g. string. The binary message is empty Expected:  Decoded to the property that contains an empty array.
- Object to encode has a property that is not contained in the schema. Expected:  Encoding fails
- Decoding a binary message that does not respect the fieldNumber order. Expected:  Decoding fails
- Decoding a binary message where one fieldNumber is used twice in the binary message. The fieldNumber order is repected, i.e., the two key-value pairs with the same fieldNumber must be adjacent in the binary message. Expected:  Decoding fails
- Encoding an object where the fieldNumbers in the schema are not sequential Expected:  Encoding succeeds
- Decoding a binary message where the fieldNumbers in the schema are not sequential Expected:  Decoding succeeds
- Decoding a binary message where one required property (not of type array) is not contained in the binary message Expected:  Decoding fails
- Decoding a binary message that contains a field number not present in the schema. Expected:  Decoding fails
- Decoding a binary message that contains some bytes at the end that cannot be decoded to some key-value pairs. E.g., a binary message contatenated with the byte 0x00. Expected:  Decoding fails
- A binary message contains a key-value pair for a property of type array for which packed encoding is used (for example boolean), and the value exist as empty bytes. Expected: Decoding fails

### Additional Information

Most of the tests should be covered by adding protocol spec.

### LINKED ISSUE #7210: Prepare Lisk SDK v6.0.0 for beta
### Description

The purpose of this epic is to prepare the Lisk SDK v6.0.0-beta.X release after fixing the bugs identified by the initial quality assurance.

### Tasks

- [x] #6361
- [x] #7215
- [x] #7265
- [x] #7090
- [x] #7235
- [x] #7261
- [x] #7239
- [x] #7260
- [x] #7047
- [x] #7156
- [x] #6201
- [x] #7996
- [x] #7993
- [x] #7953
- [x] #8012
- [x] #8007
- [x] #7799
- [x] #7994
- [x] #8006
- [x] #8037
- [x] #8021
- [x] #8058
- [x] #7150
- [x] #7010
- [x] #8038
- [x] #8039
- [x] #7969
- [x] #7989
- [x] #7614
- [x] #7320
- [x] #6627
- [x] #7301
- [x] #7108
- [x] #7376
- [x] #7258
- [x] #7143
- [x] #7227
- [x] #7213
- [x] #7139
- [x] #7232
- [x] #7225
- [x] #7252
- [x] #7253
- [x] #7254
- [x] #7255
- [x] #7276
- [x] #7140
- [x] #6710
- [x] #7299
- [x] #7328
- [x] #7332
- [x] #7291
- [x] #6864
- [x] #7338
- [x] #7309
- [x] #7391
- [x] #7393
- [x] #7473
- [x] #7437
- [x] #7466
- [x] #7005
- [x] #7372
- [x] #7387
- [x] #7349
- [x] #7380
- [x] #7330
- [x] #7189
- [x] #5317
- [x] #7149
- [x] #7277
- [x] #7172
- [x] #7229
- [x] #7236
- [x] #7231
- [x] #7272
- [x] #7251
- [x] #7331
- [x] #7333
- [x] #7311
- [x] #7321
- [x] #7305
- [x] #7334
- [x] #7353
- [x] #7355
- [x] #7362
- [x] #7807
- [x] #7834
- [x] #7800
- [x] #7783
- [x] #7631
- [x] #7563
- [x] #7512
- [x] #6990
- [x] #7526
- [x] #7528
- [x] #7632
- [x] #7570
- [x] #7611
- [x] #7625
- [x] #7465
- [x] #7467
- [x] #7469
- [x] #7493
- [x] #7462
- [x] #7506
- [x] #7297
- [x] #7598
- [x] #7542
- [x] #7447
- [x] #7603
- [x] #7601
- [x] #7620
- [x] #7635
- [x] #7660
- [x] #7568
- [x] #7636
- [x] #7710
- [x] #7738
- [x] #7739
- [x] #7802
- [x] #7808
- [x] #7692
- [x] #7812
- [x] #7762
- [x] #7837
- [x] #8009
- [x] #8010
- [x] #8011
- [x] #7972
- [x] #7983
- [x] #8013
- [x] #8014
- [x] #8015
- [x] #8017
- [x] #8018
- [x] #8016
- [x] #8019
- [x] #7982
- [x] #7981
- [x] #7980
- [x] #7979
- [x] #7978
- [x] #7947
- [x] #7948
- [x] #8008
- [x] #7944
- [x] #7956
- [x] #7987
- [x] #7946
- [x] #8026
- [x] #8042
- [x] #8050
- [x] #8052
- [x] #8053
- [x] #8060
- [x] #8057
- [x] #8063
- [x] #8070
- [x] #8072
- [x] #8074
- [x] #8080
- [x] #8088
- [x] #7678
- [x] #8033
- [x] #8034
- [x] #8035
- [x] #8077
- [x] #8091
- [x] #8041
- [x] #8028
- [x] #8083
- [x] #8085
- [x] #8069
- [x] #8082
- [x] #8061
- [x] #8022
- [x] #7002
- [x] #8056
- [x] #8075
- [x] #8086
- [x] #8190 
- [x] #8191 
- [x] #8162
- [x] #8200
- [x] #8210 
- [x] #8199
- [x] #8234
- [x] #8266
- [x] #8267
- [x] #8261
- [x] #8260
- [x] #8246
- [x] #8245
- [x] #8243
- [x] #8241
- [x] #8233
- [x] #8219
- [x] #8218
- [x] #8316 
- [x] #8318
- [x] #8319
- [x] #8303
- [x] #8291
- [x] #8290
- [x] #8289  
- [x] #8288 
- [x] #8286 
- [x] #8285
- [x] #8283  
- [x] #8282 
- [x] #8277
- [x] #8276 
- [x] #8274 
- [x] #8270 
- [x] #8268 
- [x] #8259 
- [x] #8254 
- [x] #8253 
- [x] #8252 
- [x] #8242 
- [x] #8229 
- [x] #8225 
- [x] #8224 
- [x] #8222 
- [x] #8220 
- [x] #8217 
- [x] #8215 
- [x] #8213 
- [x] #8211 
- [x] #8206
- [x] #8196 
- [x] #8194
- [x] #8188 
- [x] #8186 
- [x] #8183 
- [x] #8179 
- [x] #8177 
- [x] #8176 
- [x] #8174 
- [x] #8173 
- [x] #8172 
- [x] #8171 
- [x] #8169 
- [x] #8168 
- [x] #8165 
- [x] #8163
- [x] #8160 
- [x] #8159 
- [x] #8158 
- [x] #8157 
- [x] #8155 
- [x] #8148 
- [x] #8147 
- [x] #8145
- [x] #8141  
- [x] #8140 
- [x] #8139 
- [x] #8138 
- [x] #8132 
- [x] #8130 
- [x] #8129 
- [x] #8125 
- [x] #8124 
- [x] #8122 
- [x] #8121 
- [x] #8118 
- [x] #8114 
- [x] #8109 
- [x] #8108 
- [x] #8107 
- [x] #8102 
- [x] #8098 
- [x] #8312
- [x] #8000 
- [x] #7997 
- [x] #7974 
- [x] #8335
- [x] #8223
- [x] #8321
- [x] #8251
- [x] #7290
- [x] #8326
- [x] #8327 
- [x] #8230 
- [x] #8299 
- [x] #8334
- [x] #8330 
- [x] #8347 
- [x] #8151 
- [x] #8232 
- [x] #8301 
- [x] #8338 
- [x] #8342 
- [x] #8184 
- [x] #8382 
- [x] #8379
- [x] #8376
- [x] #8373
- [x] #8366
- [x] #8364
- [x] #8359 
- [x] #8357
- [x] #8356 
- [x] #8355
- [x] #8351
- [x] #8339  
- [x] #8349

### PR DESCRIPTION
**Pull request by mitsuaki-u**
### What was the problem?

This PR resolves #7047

### How was it solved?

- Update codec

### How was it tested?

- Add unit tests

### CODE REVIEW

File: elements/lisk-p2p/src/utils/codec.ts [hunk 1]
```diff
@@ -44,4 +44,7 @@ export const encodePeerInfo = (peerInfoSchema: Schema, data: ProtocolPeerInfo):
 	codec.encode(peerInfoSchema, data);
 
 export const encodeNodeInfo = (nodeInfoSchema: Schema, data: P2PNodeInfo): Buffer =>
-	codec.encode(nodeInfoSchema, data);
+	codec.encode(nodeInfoSchema, {
+		...data,
+		advertiseAddress: data.advertiseAddress ?? false,
```
Comment (shuse2) [inline elements/lisk-p2p/src/utils/codec.ts:49]: Here, if advertiseAddress is not defined, encode `false` because no optional field is allowed

**Review by shuse2 — APPROVED** [review]

File: elements/lisk-codec/test/decode.spec.ts [hunk 2]
```diff
@@ -129,6 +149,180 @@ describe('decode', () => {
...
+			const keypair1 = Buffer.concat([key, intVal1]);
+			const keypair2 = Buffer.concat([key2, intVal2]);
+			const encodedValue = Buffer.concat([keypair2, keypair1]);
+
+			expect(() => codec.decode(schema, encodedValue)).toThrow(
+				'Invalid field number while decoding',
+			);
+		});
+
+		it('should not decode a binary message that contains some bytes at the end that cannot be decoded to some key-value pairs. E.g., a binary message contatenated with the byte 0x00', () => {
```
Comment (ishantiw) [inline elements/lisk-codec/test/decode.spec.ts:280]:
```suggestion
		it('should not decode a binary message that contains some bytes at the end that cannot be decoded to some key-value pairs. E.g., a binary message concatenated with the byte 0x00', () => {
```

File: elements/lisk-codec/test/decode.spec.ts [hunk 3]
```diff
@@ -129,6 +149,180 @@ describe('decode', () => {
...
+
+			const keypair1 = Buffer.concat([key, intVal1]);
+			const encodedValue = Buffer.concat([keypair1, Buffer.alloc(20, 255)]);
+
+			expect(() => codec.decode(schema, encodedValue)).toThrow('Invalid terminate index.');
+		});
+
+		it('should not decode a binary message which contains a key-value pair for a property of type array for which packed encoding is used (for example boolean), and the value exist as empty bytes', () => {
+			const schema = {
+				$id: 'duplicate/emptyBytes',
```
Comment (ishantiw) [inline elements/lisk-codec/test/decode.spec.ts:302]: This schema also includes non-sequential fieldNumber, maybe to only have empty bytes case

File: elements/lisk-codec/test/decode.spec.ts [hunk 4]
```diff
@@ -129,6 +149,180 @@ describe('decode', () => {
...
+			const key = generateKey(schema.properties.data);
+			const intVal1 = writeUInt32(20);
+
+			const keypair1 = Buffer.concat([key, intVal1]);
+			const encodedValue = Buffer.concat([keypair1, Buffer.alloc(20, 255)]);
+
+			expect(() => codec.decode(schema, encodedValue)).toThrow('Invalid terminate index.');
+		});
+
+		it('should not decode a binary message which contains a key-value pair for a property of type array for which packed encoding is used (for example boolean), and the value exist as empty bytes', () => {
```
Comment (ishantiw) [inline elements/lisk-codec/test/decode.spec.ts:300]: Maybe my understanding is incorrect but `contains a key-value pair for a property of type array` is it an array or an object?

**Review by ishantiw — COMMENTED** [review]
LGTM, just a few comments

File: elements/lisk-codec/test/decode.spec.ts (hunk 4 above)
Comment (shuse2) [reply elements/lisk-codec/test/decode.spec.ts:300]: The testing schema was wrong. the `an array` is correct. the encoding always has key-value pair indicating `fieldNumber/propertyType => bytes` , and it means only the key exist, but the value is empty then it should fail

**Review by ishantiw — APPROVED** [review]
🚀

**Pull request merged**